### DISCUSSION TITLE
Prompt Flexibility

### DISCUSSION THREAD

**Discussion by RandomCleverName** [Kilvoctu/aiyabot#135] [Ideas]
The issue I'm having/trying to solve is in bold for anyone who just wants to get to the meat of it.

I've been toying around adding and subtracting things that I don't feel is the fit for me or need in my bot (like hyperlinks, loras, etc). I've done pretty good up until now... I've run into a road block and if someone wants to point me in the right direction or help in anyway. it'd be appreciated.

So my issue... **The config file has a negative prompt prefix. Since with this kinda stuff it matters if its prefix or suffix and I've been trying, and failing to add a suffix. I've also been trying to add a positive prompt suffix & prefix to no avail. Just been opening files looking for anything and everything that references prompts to try and at least find where the edit should happen**

This is a hobby for me, so it may be somewhere obvious, but I've just been opening things in notepad++ and scanning for what references it that keeps popping the mismatch and rewriting it.

Anyone point me in the right direction on this? or is anyone already working on it/done with it and I'm just wasting my time?

**Comment by Kilvoctu** [comment]
Heya, the prompt moderation stuff in stablecog.py and viewhandler.py is running a function (`prompt_mod`) that sits in settings.py.

https://github.com/Kilvoctu/aiyabot/blob/15d41f3d7e1962b768650144bac149c0329edfac/core/settings.py#L133-L157
On line 155, this is where the words in `negative_prompt_prefix` (as `z`) are added to the beginning of the negative prompt.  
You can edit this function to do prompt prefix/suffix, too. 

I do this as a hobby, too. I'm no programmer, no knowledge of good practices or fundamentals, so a lot of the code is "hm this feels right, and it works".
It makes sense to me, but probably not as much to others. If you have other questions or need help, feel free to reply  👍👍

**Comment by RandomCleverName** [comment]
Well, my trial and error skills and google skills have pretty much broken for the day. I got as far as changing the name before I broke it pretty bad trying to move it to a suffix, glad to have had backups lol. Move the Z and the Y, didn't break anything but didn't work either. So now I'm scanning other files to find anything I need to move as well.

Prolly something really simple I'm over looking and will notice if I come back bright eyed and bushy tailed. Should be easier than adding a whole prompt suffix I would imagine... Really though that may be out of my depth but it's not gonna stop me from trying... Well, least til my wife gets mad at me for wasting a week off in front of a PC. TBH if that happens at the rate I'm going and the rate I get distracted I may just adopt a "if it ain't broke don't fix it" attitude.

**Reply by Kilvoctu** [reply]
On line 155, only replacing `negative_prompt = f"{z} {negative_prompt}"` with `negative_prompt = f"{negative_prompt} {z}"` worked for me.

The exception was if a Web UI style was added, since this is added afterward (during generation). I got around this by using a lesser known [feature](https://github.com/AUTOMATIC1111/stable-diffusion-webui/wiki/Features#styles) of Web UI. Adding `{prompt}` to a style inserts the user's prompt at that location.  
So if your Web UI negative prompt style has `1, 2, {prompt}`, and user's negative prompt is `3`, the result will be `1, 2, 3`.  
If the Web UI negative prompt style is instead `1, 2`, you'll get `3, 1, 2`.

Also, if you're making changes and experimenting, I'd recommend forking the repo. You can have your main branch that works, and other branches for testing changes.  
The reason why I have so many [PR](https://github.com/Kilvoctu/aiyabot/pulls?q=is%3Apr+is%3Aclosed) for the aiyabot is because trial and error is how I do things too. I test and break things on new branches, then merge into main when it's working.

**Reply by RandomCleverName** [reply]
Sorry, just woke up. Was sitting here with my cup of coffee looking at it again and refreshed this tab and now it makes total sense... That being said the `{prompt}` tip is something I didn't know, very helpful indeed, Armed with that knowledge I should be able to knock this out pretty quick. Turns out yesterday I was moving the wrong `Z` at `153` even though I know for a fact you were talking about `155`. Think I just sat here so long copying the AI files and building it be extra NSFW to my external yesterday that my brain crapped out. That tip on `{prompt}` placing being positional really does save a lot of time, I genuinely had no idea and thank and appreciate you taking your time to deal with me, cheers mate ☕ .

I gave up and went to bed yesterday when the bot broke down and required me to transfer it directly to C:\ to make it work again, no idea why it did that, but at least with that fixed and your tip now I can stalk my way back to figuring out why the bot outputs themselves bypass the NSFW filter extension. Edit/Findings on this subject: After reading your other response to the other thread I;m wondering if its possible. I see you suggest just writing the script into the base itself which I may give a whirl. Also just a heads up on the banned word list, I know you said you could type partial words, but is defeated by "leet speak" as my wife figured out.

Then on to figuring out why 1 of my models wont show up (has an ë which is suspect is the cause), and then on to my last project of finishing my home automation (this serves as a break from that, but should be the other way around with how long the bot takes me in comparison 😆).

Thanks again for the help on this issue!

**Reply by Kilvoctu** [reply]
No problem, I can try to help if you have any other questions.

The aiyabot (and other frontends) works by talking through the API supplied by Web UI, so we're limited by what this allows. So as you deduced, Web UI extensions and scripts don't work unless specifically instructed to do so through API (if one exists for the extension/script). Web UI scripts have gotten better support recently, but I've not had any interest in looking into it.

Also yeah, the banned word implementation is really quickly done; I know how easily it's defeated, lol. I think I said in the other thread it's there so there's something rather than nothing.

> Then on to figuring out why 1 of my models wont show up (has an ë which is suspect is the cause)

This [project](https://github.com/users/Kilvoctu/projects/1/views/1) is supposed to solve model configuration woes, but it keeps getting put off for months. Not enough time to do everything, and I usually swap priority to what people are requesting.  
I think in this case it's an encoding issue, I'll look at it.  
edit: I renamed one of my model to have an "ë" in it, and it did appear on my AIYA, so it may be something else.  
![image](https://user-images.githubusercontent.com/2993060/222974435-d27f0537-790d-46d3-8980-55247364a9a0.png)

**Reply by RandomCleverName** [reply]
Understandable, thought I was making head way on having the output picture checked using open_NSFW, then broken things so went to another. Got close again using NudeNet, then I broke everything, and by everything I do mean everything lol. Now I'm trying to approach it differently now. If I can get the sent output to come from the actual folder, where things are blacked out, I think that I'd be in good shape. Easier said than done though. Not even sure how to approach such a thing off the top of my head, so brain storm and google time. 

Interesting about the ë, I assumed it was the issue but I suppose back to the drawing board on that as well, for now though time to visit mother in-law so it'll be on my back burner for a bit.
